### PR TITLE
[Improvement-4450][datasource] Hive/Spark data sources do not support multi-tenancy when Kerberos authentication is enabled

### LINKED ISSUE #4450: [Bug][datasource] Hive/Spark data sources do not support multi-tenancy when Kerberos authentication is enabled
**Describe the bug**
Hive/Spark JDBC supports manual configuration of keytab parameters when using Kerberos authentication, and supports multi-tenancy when using Kerberos authentication for Hive/Spark data sources.

The current Hive/Spark JDBC system uses Kerberos authentication using the keytab configured in the common.properties file. There is a problem when Hive/Spark data source Kerberos authenticates other users or multiple different users.

**To Reproduce**
1.common.properties configuration keytab file is for the DolphinScheduler tenant;
2.Hive DataSource Kerberos authenticated tenants are hive_exec_1 tenants;
3.Use the above Hive data source query to report no permissions;

**Which version of Dolphin Scheduler:**
 -[dev]

**Requirement or improvement**
Improvement: Add Kerberos-related parameters to the other parameters or principal parameters section of the DataSource page for manual input by the user

**Additional context**
Points needing attention:
Kerberos expiration issue (I don't think this Kerberos expiration is related to scheduling, we will consider whether to deal with this later)

Solution 1:
Execute the script kinit-kt/XXX /xxx.keytab xxx@xxx.xxx.com using the shell task node, and set the keytab to be swipe regularly

If you have any good ideas, please leave a message!

---

**描述错误**
Hive/Spark JDBC 使用Kerberos认证时支持keytab相关参数手动配置，解决Hive/Spark数据源开启Kerberos认证时支持多租户。

当前系统Hive/Spark JDBC 使用Kerberos认证时使用的都是Common.properties文件内配置的keytab,当Hive/Spark数据源kerberos认证其他用户或多个不同的用户是存在问题的。

**To Reproduce**
1.common.properties 配置 keytab文件 是dolphinscheduler租户的；
2.hive datasource kerberos认证租户是 hive_exec_1租户的；
3.使用上面的hive数据源查询语句会报没有权限；

**要求或改进**
改进：在datasource页面其他参数或principal参数部分新增kerberos相关参数，供用户手动输入

**Additional context**
需要注意的点：
kerberos过期问题（我认为这个kerberos过期和调度无关，这块后面在考虑要不要处理）

解决方案一：
使用shell任务节点执行 kinit-kt/XXX /xxx.keytab xxx@xxx.xxx.com 脚本，设置定时刷keytab

你有什么好的想法，欢迎留言！

### PR DESCRIPTION
**Pull request by zhuangchong**
## What is the purpose of the pull request
this PR closes #4450 

Hive/Spark JDBC supports manual configuration of keytab parameters when using Kerberos authentication, and supports multi-tenancy when using Kerberos authentication for Hive/Spark data sources.

The current Hive/Spark JDBC system uses Kerberos authentication using the keytab configured in the common.properties file. There is a problem when Hive/Spark data source Kerberos authenticates other users or multiple different users.

iShot2021-01-27 20 31 59

iShot2021-01-27 20 47 27

iShot2021-01-27 20 52 18

## Verify this pull request
This change added tests and can be verified as follows:
  - *Manually verified the change by testing locally.*

### CODE REVIEW

**Comment by zhuangchong** [comment]
@dailidong  help review

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/DataSourceController.java [hunk 1]
```diff
@@ -152,7 +160,11 @@ public Result createDataSource(@ApiIgnore @RequestAttribute(value = Constants.SE
             @ApiImplicitParam(name = "userName", value = "USER_NAME", required = true, dataType = "String"),
             @ApiImplicitParam(name = "password", value = "PASSWORD", dataType = "String"),
             @ApiImplicitParam(name = "connectType", value = "CONNECT_TYPE", dataType = "DbConnectType"),
-            @ApiImplicitParam(name = "other", value = "DATA_SOURCE_OTHER", dataType = "String")
+            @ApiImplicitParam(name = "other", value = "DATA_SOURCE_OTHER", dataType = "String"),
+            @ApiImplicitParam(name = "principal", value = "DATA_SOURCE_PRINCIPAL", dataType = "String"),
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/DataSourceController.java:164]: The order of the parameters seems a bit problematic

**Comment by CalvinKirs** [comment]
PTAL @break60

**Comment by break60** [comment]
The three newly added fields are required fields, and the front-end has not done a required check

**Comment by zhuangchong** [comment]
> The three newly added fields are required fields, and the front-end has not done a required check

The three newly added fields are non-required fields

**Comment by break60** [comment]
> > The three newly added fields are required fields, and the front-end has not done a required check
> 
> The three newly added fields are non-required fields

image
If it is not required, please remove the front *

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/DataSourceController.java (hunk 1 above)
Comment (zhuangchong) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/DataSourceController.java]: done

**Comment by zhuangchong** [comment]
> > > The three newly added fields are required fields, and the front-end has not done a required check
> > 
> > 
> > The three newly added fields are non-required fields
> 
> image
> If it is not required, please remove the front *
> image

done

**Comment by break60** [comment]
> > > > The three newly added fields are required fields, and the front-end has not done a required check
> > > 
> > > 
> > > The three newly added fields are non-required fields
> > 
> > 
> > image
> > If it is not required, please remove the front *
> > image
> 
> done

Please check the error on line 384 of the createDataSource.vue file

**Review by CalvinKirs — APPROVED** [review]
LGTM.(backend)

**Review by break60 — APPROVED** [review]
front end +1

**Comment by CalvinKirs** [comment]
@CalvinKirs Can 1.3.6 be released?

**Comment by geosmart** [comment]
when a tenant's user add datasource, the keytab file path is locate in api server ? or user need to manualy upload a keytab file.


maybe we can upload user's keytab file with username.keytab to a hdfs path with admin user;

and as default ,user load keytab file by $username.keytab to get a GUI to doAs(operation).

so user doesn't need to config krb5 with every hive datasource

**Comment by geosmart** [comment]
in some group, the user doesn't know where the keytab file store, and just know  the ldap password to login.

and for security, the keytab file should not upload to platform by user

**Pull request merged**